hi-ml/testhiml/testhiml/test_run_ml.py

```diff
@@ -204,19 +252,78 @@ def _expected_files_exist() -> bool:
+                            mock_run_inference.assert_called_once()
+
+
+def test_run_inference_only(ml_runner_with_run_id: MLRunner) -> None:
+    ml_runner_with_run_id.container.run_inference_only = True
+    assert ml_runner_with_run_id.checkpoint_handler.trained_weights_path
+    with patch.object(ml_runner_with_run_id, "run_training") as mock_run_training:
+        with patch.object(ml_runner_with_run_id, "run_validation") as mock_run_validation:
+            with patch.object(ml_runner_with_run_id, "run_inference") as mock_run_inference:
+                ml_runner_with_run_id.run()
+                mock_run_training.assert_not_called()
+                mock_run_validation.assert_not_called()
+                mock_run_inference.assert_called_once()
 
-                ml_runner_with_container.run()
 
-                mock_load.assert_called_once()
-                assert ml_runner_with_container._has_setup_run
-                assert ml_runner_with_container.checkpoint_handler.has_continued_training
+@pytest.mark.parametrize("run_extra_val_epoch", [True, False])
+def test_resume_training_from_run_id(run_extra_val_epoch: bool, ml_runner_with_run_id: MLRunner) -> None:
+    ml_runner_with_run_id.container.run_extra_val_epoch = run_extra_val_epoch
+    ml_runner_with_run_id.container.max_epochs = 2
+    ml_runner_with_run_id.container.max_num_gpus = 0
+    assert ml_runner_with_run_id.checkpoint_handler.trained_weights_path
+    with patch.object(ml_runner_with_run_id, "run_validation") as mock_run_validation:
+        with patch.object(ml_runner_with_run_id, "run_inference") as mock_run_inference:
+            ml_runner_with_run_id.run()
+            assert mock_run_validation.called == run_extra_val_epoch
+            mock_run_inference.assert_called_once()
+
+
+@pytest.mark.parametrize("run_inference_only", [True, False])
+def test_load_model_checkpoint(run_inference_only: bool, mock_run_id: str, tmp_path: Path) -> None:
+    experiment_config = ExperimentConfig(model="HelloWorld")
+    container = HelloWorld()
+    container.max_num_gpus = 0
+    container.save_checkpoint = True
+    container.ckpt_run_id = mock_run_id
+    container.run_inference_only = run_inference_only
+    container.run_extra_val_epoch = True
+    with patch("health_azure.utils.get_workspace") as mock_get_workspace:
+        mock_get_workspace.return_value = _get_workspace(tmp_path)
+        runner = MLRunner(experiment_config=experiment_config, container=container)
+        runner.setup()
+        weights_before: torch.Tensor = container.model.model.weight.detach().clone()  # type: ignore
+        with patch.object(runner, "run_validation"):
+            with patch.object(runner, "run_inference"):
+                runner.run()
+        weights_after: torch.Tensor = runner.container.model.model.weight  # type: ignore
+        assert not torch.allclose(weights_before, weights_after)
```

Comment:
I'm not convinced we're asserting the right thing here, given the test name. It makes sense when `run_inference_only=True` that the only way that the weights have been updated is because `load_model_checkpoint` was called - but when `run_inference_only=False` then training occurs so the weights will update regardless of whether `load_model_checkpoint` gets called. Does it make more sense to mock `load_model_checkpoint` and assert that it has been called?

Reply:
`load_model_checkpoint`[ is not called systematically for all containers](https://github.com/microsoft/hi-ml/blob/b66b42c7ed4c67b4b6b385b0bdf5862dcedde92d/hi-ml/src/health_ml/run_ml.py#L337). if `test_step `is not specified, we don't run inference nor load the model checkpoint unless we activate the extra validation epoch...
 
The HelloRegression Lightning Module does not override `test_step `so `load_model_checkpoint` won't be called by default. 
the container.model weights will neverthless change by the trainer but won't necessarily correspond to the "best checkpoint" (it's usually the last checkpoint by default)

Here I just want to make sure that we're using the right weights in 2 scenarios:
run_inference_only=True: all training will be skipped, and I want to make sure that inference is run with the loaded checkpoint (weights_before != weights_after) and here `load_model_checkpoint `was indeed called
run_inference_only=False: but a ckpt_run_id is provided so we want to resume training from there, however I can't really test that because it's passed to the trainer through `resume_from_checkpoint` and that would be testing external code... however I still want to make sure that the weights have changed at this point, here `load_model_checkpoint ` is not called...

Does it make sense? Happy to follow up offline